neo4j/internal/racing/lock.go

```diff
@@ -30,14 +30,16 @@ func NewMutex() Mutex {
 }
 
 func (c *contextLock) TryLock(ctx context.Context) bool {
-	if deadline, hasDeadline := ctx.Deadline(); !hasDeadline {
+	deadline, hasDeadline := ctx.Deadline()
+	err := ctx.Err()
+	switch {
+	case !hasDeadline && err == nil:
 		c.ch <- struct{}{}
 		return true
-	} else if deadline.Before(time.Now()) {
-		// if the context is already done and the lock is available
-		// select will pick any of the case at random
+	case deadline.Before(time.Now()) || err != nil:
```

Comment:
Is there a potential race-condition? What if, at this point, the deadline has not yet expired, but it does between this check the the following `select` block? I'm wondering, is the `.Before` check even necessary? What if the user wants to try to acquire the lock without waiting (timeout 0 seconds). What context would they chose?

This might be controversial but I'd suggest that on expired contexts, we try to acquire the lock and if we can't do so immediately, we return `false`. Or in other words a non-blocking lock acquisition.

Reply:
To elaborate: the sole purpose of this early check is to avoid the situation where the lock is available and the context has expired at the same time. In that specific case, without this early check, `select` would pick at random, which is not what we want. If both the lock is available and the context has expired, we don't want to acquire the lock.

If this check passes, it does not necessarily mean the lock will be acquired, as you pointed out. But I don't think it's an issue.

Reply:
Fair enough. But the random choice can still happen iff the first check passes and the context times out right afterwards.

Reply:
true, but the window becomes much smaller so I think it's acceptable